spec/lib/cache_spec.rb

```diff
@@ -104,4 +102,27 @@
 
     expect(cache.redis.ttl("#{cache.namespace}:foo:bar")).to eq(180)
   end
+
+  describe ".fetch" do
+    subject(:fetch_value) { cache.fetch(key) { "bob" } }
+
+    let(:key) { "my_key" }
+
+    before { cache.delete(key) }
+
+    context "when the cache is corrupt" do
+      let(:redis_key) { cache.normalize_key(key) }
+
+      before { Discourse.redis.setex(redis_key, described_class::MAX_CACHE_AGE, "") }
+
+      it "runs and return the provided block" do
+        expect(fetch_value).to eq("bob")
+      end
+
+      it "generates a new cache entry" do
+        fetch_value
+        expect(cache.read(key)).to eq "bob"
+      end
+    end
```

Comment:
plz plz plz, can we make this more concise? 😅 

```suggestion
    context "when the cache is corrupt" do
      before do
        cache.delete("my_key")
        redis_key = cache.normalize_key("my_key")
        Discourse.redis.setex(redis_key, described_class::MAX_CACHE_AGE, "")
      end

      it "runs and return the provided block" do
        expect(cache.fetch("my_key") { "bob" }).to eq("bob")
      end

      it "generates a new cache entry" do
        cache.fetch("my_key") { "bob" }
        expect(cache.read("my_key")).to eq("bob")
      end
    end
```

consider this test:

```rb
it "generates a new cache entry" do
  cache.fetch("my_key") { "bob" }
  expect(cache.read("my_key")).to eq("bob")
end
```

step 1: save value to `cache`
step 2: read the value from `cache`
only one external test dependency, `cache`

Instead of this (that's how I read "everything is a `let`" rspec code 😅)

<img width="727" alt="Screenshot 2024-06-18 at 12 53 46" src="https://github.com/discourse/discourse/assets/66961/c697ff59-c165-4370-993c-d7fe288f2a87">

Reply:
I can probably remove some stuff yes, but otherwise I’m really used to writing “proper” specs, following the existing style guides and conventions. I understand some people don’t like RSpec (like DHH), but we don’t have clear guidelines on writing specs. Sometimes, it really feels like we should just use `TestCase` instead of specs (so you can do pretty much whatever you want with it :joy:).

The idea behind putting `before` blocks (or anything else) in their proper `context`/`describe` block, is to then facilitate the extension of the spec for example.

Reply:
To me all this structure does is obfuscating the code.

> The idea behind putting before blocks (or anything else) in their proper context/describe block, is to then facilitate the extension of the spec for example.

This is "premature optimization" of tests. If somebody (unlikely) wants to add a spec that uses a different key they can extract it, or just write `cache.fetch("my_other_key")` directly. There's no need to preemptively chop it up into extendable pieces 😅

Reply:
> To me all this structure does is obfuscating the code.

Well, I guess it can be perceived like this. The idea in theory being to put all the setup code outside the example and help focussing on what’s really under test. I’m probably too biased now anyway :sweat_smile: 

> This is "premature optimization" of tests.

Regarding `let`s , yes you’re probably right :grin: I’ll update the spec :wink: